# GWAS
GWAS is Genome-wide association studies to identify genotype and phenotype's association, especially single nucleotide polymorphisms (SNPs). GWAS can be used to explore the genetice risk for individuals.
here is the tutorial for GWAS analysis: https://pmc.ncbi.nlm.nih.gov/articles/PMC6001694/
Marees, A. T., de Kluiver, H., Stringer, S., Vorspan, F., Curis, E., Marie-Claire, C., & Derks, E. M. (2018). A tutorial on conducting genome-wide association studies: Quality control and statistical analysis. International journal of methods in psychiatric research, 27(2), e1608. https://doi.org/10.1002/mpr.1608

In this chapter, I will put more details and illustration to make it easier to understand and operate for beginners.

the major software used is PLINK. https://www.cog-genomics.org/plink2/
```bash
# use appropraite link for system, mine is linux 64-bit
wget https://s3.amazonaws.com/plink1-assets/plink_linux_x86_64_20241022.zip
unzip plink_linux_x86_64_20241022.zip
./plink

## add to path
echo 'export PATH=$PATH:path/to/plink' >> ~/.bashrc
source ~/.bashrc

## check
plink --version

## or locate into directory
./plink
```

Quality control steps are essential to clean GWAS data and make its findings more statistically meaningful, the paper provide detailed codes from QC to analyze

## data obtaining
the reference paper using HapMap database, while this database is out of dated in 2025, for data acquiring of GWAS, here is one database: https://gwas.mrcieu.ac.uk/. For example, searching autism, and select database: ieu-a-1185, which will provide VCF data and VCF index. From here, plink can be used to build bed, bim and fam files.

```bash
# get data
wget https://gwas.mrcieu.ac.uk/files/ieu-a-1185/ieu-a-1185.vcf.gz
wget https://gwas.mrcieu.ac.uk/files/ieu-a-1185/ieu-a-1185.vcf.gz.tbi

# processed by plink
plink --vcf ieu-a-1185.vcf.gz --make-bed --out ASD_GWAS_data
plink --bfile ASD_GWAS_QC --missing --out missing_data
```
back to reference paper 
```bash
# HapMap_3_r3_1 is the main name of three files HapMap_3_r3_1.bed, HapMap_3_r3_1.bim,and HapMap_3_r3_1.fam
# bim file contains genome coordinates of SNPs, and allel information
head HapMap_3_r3_1.bim 
1	rs2185539	0	556738	T	C
# rs2185539 snp locates in chr1 position 556738 with reference allel C and alternative allel T

# fam file contains sample information 
head HapMap_3_r3_1.fam 
1328 NA06989 0 0 2 2
# family ID is 1328 and sample ID is NA06989 father id is 0, mother id is 0, sex is female (1 for male), and phenotype is case (1 for control, -9 for missing)
# the description of data format in chapter 2.1 of reference paper.

# check the missing value per individual(.imiss) and SNP (.lmiss)
plink --bfile HapMap_3_r3_1 --missing

## filter out some missingness
# Delete SNPs with missingness >0.2.
plink --bfile HapMap_3_r3_1 --geno 0.2 --make-bed --out HapMap_3_r3_2

# Delete individuals with missingness >0.2.
plink --bfile HapMap_3_r3_2 --mind 0.2 --make-bed --out HapMap_3_r3_3

# Delete SNPs with missingness >0.02.
plink --bfile HapMap_3_r3_3 --geno 0.02 --make-bed --out HapMap_3_r3_4

# Delete individuals with missingness >0.02.
plink --bfile HapMap_3_r3_4 --mind 0.02 --make-bed --out HapMap_3_r3_5

# check sex, obtained (.sexcheck file)
plink --bfile HapMap_3_r3_5 --check-sex

# estimate sex and remake the data
plink --bfile HapMap_3_r3_5 --impute-sex --make-bed --out HapMap_3_r3_6

# Select autosomal SNPs only (i.e., from chromosomes 1 to 22).
awk '{ if ($1 >= 1 && $1 <= 22) print $2 }' HapMap_3_r3_6.bim > snp_1_22.txt
plink --bfile HapMap_3_r3_6 --extract snp_1_22.txt --make-bed --out HapMap_3_r3_7

# Generate a plot of the MAF distribution.
plink --bfile HapMap_3_r3_7 --freq --out MAF_check

# Remove SNPs with a low MAF frequency.
plink --bfile HapMap_3_r3_7 --maf 0.05 --make-bed --out HapMap_3_r3_8
# 1073226 SNPs are left

# Delete SNPs which are not in Hardy-Weinberg equilibrium (HWE).
# Check the distribution of HWE p-values of all SNPs.
plink --bfile HapMap_3_r3_8 --hardy

# Selecting SNPs with HWE p-value below 0.00001, required for one of the two plot generated by the next Rscript, allows to zoom in on strongly deviating SNPs. 
awk '{ if ($9 <0.00001) print $0 }' plink.hwe>plinkzoomhwe.hwe

# By default the --hwe option in plink only filters for controls.
# Therefore, we use two steps, first we use a stringent HWE threshold for controls, followed by a less stringent threshold for the case data.
plink --bfile HapMap_3_r3_8 --hwe 1e-6 --make-bed --out HapMap_hwe_filter_step1

# The HWE threshold for the cases filters out only SNPs which deviate extremely from HWE. 
# This second HWE step only focusses on cases because in the controls all SNPs with a HWE p-value < hwe 1e-6 were already removed
plink --bfile HapMap_hwe_filter_step1 --hwe 1e-10 --hwe-all --make-bed --out HapMap_3_r3_9

# Theoretical background for this step is given in our accompanying article: https://www.ncbi.nlm.nih.gov/pubmed/29484742 .

# The parameters ë50 5 0.2í stand respectively for: the window size, the number of SNPs to shift the window at each step, and the multiple correlation coefficient for a SNP being regressed on all other SNPs simultaneously.

plink --bfile HapMap_3_r3_9 --exclude inversion.txt --range --indep-pairwise 50 5 0.2 --out indepSNP
# Note, don't delete the file indepSNP.prune.in, we will use this file in later steps of the tutorial.

plink --bfile HapMap_3_r3_9 --extract indepSNP.prune.in --het --out R_check
# This file contains your pruned data set.
# For data manipulation we recommend using UNIX. However, when performing statistical calculations R might be more convenient, hence the use of the Rscript for this step:
Rscript --no-save heterozygosity_outliers_list.R

# Adapt this file to make it compatible for PLINK, by removing all quotation marks from the file and selecting only the first two columns.
sed 's/"// g' fail-het-qc.txt | awk '{print$1, $2}'> het_fail_ind.txt

# Remove heterozygosity rate outliers.
plink --bfile HapMap_3_r3_9 --remove het_fail_ind.txt --make-bed --out HapMap_3_r3_10

# Check for relationships between individuals with a pihat > 0.2.
plink --bfile HapMap_3_r3_10 --extract indepSNP.prune.in --genome --min 0.2 --out pihat_min0.2


# The HapMap dataset is known to contain parent-offspring relations. 
# The following commands will visualize specifically these parent-offspring relations, using the z values. 
awk '{ if ($8 >0.9) print $0 }' pihat_min0.2.genome>zoom_pihat.genome

# To demonstrate that the majority of the relatedness was due to parent-offspring we only include founders (individuals without parents in the dataset).

plink --bfile HapMap_3_r3_10 --filter-founders --make-bed --out HapMap_3_r3_11

# Now we will look again for individuals with a pihat >0.2.
plink --bfile HapMap_3_r3_11 --extract indepSNP.prune.in --genome --min 0.2 --out pihat_min0.2_in_founders


# For each pair of 'related' individuals with a pihat > 0.2, we recommend to remove the individual with the lowest call rate. 
plink --bfile HapMap_3_r3_11 --missing


```


